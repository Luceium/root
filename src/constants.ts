export var STATUS = {
  INCOMPLETE: "incomplete",
  SUBMITTED: "submitted",
  WAITLISTED: "waitlisted",
  REJECTED: "rejected",
  ADMITTED: "admitted",
  ADMISSION_CONFIRMED: "admission_confirmed",
  ADMISSION_DECLINED: "admission_declined"
}

export var TYPE = {
  IN_STATE: "is",
  OUT_OF_STATE: "oos",
  STANFORD: "stanford"
}

export var DEADLINES = [
  {
    "key": "oos",
    "label": "out-of-state",
    "date": "2018-11-20T07:59:00.000Z",
    "display_date": "November 19, 2018"
  },
  {
    "key": "is",
    "label": "in-state",
    "date": "2018-11-27T07:59:00.000Z",
    "display_date": "November 26, 2018"
  },
  {
    "key": "stanford",
    "label": "Stanford student",
    "date": "2019-02-14T07:59:00.000Z",
    "display_date": "February 13, 2019"
  }
];

export var TRANSPORTATION_STATUS = {
  PENDING: "pending",
  UNAVAILABLE: "unavailable",
  AVAILABLE: "available",
  SUBMITTED: "submitted",
  REJECTED: "rejected",
  APPROVED: "approved",
  PAID: "paid"
};

export var TRANSPORTATION_TYPES = {
  BUS: "bus",
  FLIGHT: "flight",
  OTHER: "other"
};

export var TRANSPORTATION_BUS_ROUTES = {
  TEST: "test",
  USC: "usc",
  UCLA: "ucla",
  SANDIEGO: "sandiego",
  UCI: "uci",
  POMONA: "pomona",
  BERKELEY: "berkeley"
};

export var TRANSPORTATION_DEADLINES = {
  [TRANSPORTATION_TYPES.FLIGHT]: 'December 9th at 11:59pm PST',
  [TRANSPORTATION_TYPES.BUS]: 'January 10th at 11:59pm PST',
  [TRANSPORTATION_TYPES.OTHER]: 'February 19th at 11:59pm PST'
};


export type IBusRoute = {
  day?: string,
  time?: string,
  stop?: string,
  location?: string,
  hack?: boolean
};

export var TRANSPORTATION_BUS_ROUTE_DETAILS: { [x: string]: IBusRoute[] } = {
  [TRANSPORTATION_BUS_ROUTES.USC]: [
    {
      day: 'Friday, February 15th',
      time: '8:00am',
      stop: 'Check in at 37th & McClintock',
      location: 'Los Angeles'
    },
    {
      day: 'Friday, February 15th',
      time: '4:30pm',
      stop: 'Panama St. & Via Ortega',
      location: 'Stanford'
    },
    {
      hack: true
    },
    {
      day: 'Sunday, February 17th',
      time: '4:00pm',
      stop: 'Check in at Panama St. & Via Ortega',
      location: 'Stanford'
    },
    {
      day: 'Sunday, February 17th',
      time: '11:00pm',
      stop: '37th & McClintock',
      location: 'Los Angeles'
    }
  ],
  [TRANSPORTATION_BUS_ROUTES.TEST]: [
    {
      day: 'Friday, February 15th',
      time: '8:00am',
      stop: 'Check in at 37th & McClintock',
      location: 'Atlanta'
    },
    {
      hack: true
    },
    {
      day: 'Sunday, February 17th',
      time: '11:00pm',
      stop: '37th & McClintock',
      location: 'Los Angeles'
    }
  ],
  [TRANSPORTATION_BUS_ROUTES.TEST]: [
    {
      day: 'Friday, February 15th',
      time: '9:00am',
      stop: 'Check in at 10 Charles E. Young',
      location: 'Los Angeles'
    },

    {
      day: 'Friday, February 15th',
      time: '4:30pm',
      stop: 'Panama St. & Via Ortega',
      location: 'Stanford'
    },

    {
      hack: true
    },

    {
      day: 'Sunday, February 17th',
      time: '4:00pm',
      stop: 'Check in at Panama St. & Via Ortega',
      location: 'Stanford'
    },

    {
      day: 'Sunday, February 17th',
      time: '10:30pm',
      stop: '10 Charles E. Young',
      location: 'Los Angeles'
    }
  ],
  [TRANSPORTATION_BUS_ROUTES.SANDIEGO]: [
    {
      day: 'Friday, February 15th',
      time: '7:30am',
      stop: 'Check in at UCSD (location added later)',
      location: 'Los Angeles'
    },

    {
      day: 'Friday, February 15th',
      time: '6:00pm',
      stop: 'Panama St. & Via Ortega',
      location: 'Stanford'
    },

    {
      hack: true
    },

    {
      day: 'Sunday, February 17th',
      time: '3:30pm',
      stop: 'Check in at Panama St. & Via Ortega',
      location: 'Stanford'
    },

    {
      day: 'Sunday, February 17th',
      time: '11:45pm',
      stop: 'UCSD',
      location: 'Los Angeles'
    }
  ],
  [TRANSPORTATION_BUS_ROUTES.UCI]: [
    {
      day: 'Friday, Februrary 15th',
      time: '8:45am',
      stop: 'Check in at UCI (location added later)',
      location: 'Irvine'
    },

    {
      day: 'Friday, Feburary 15th',
      time: '6:00pm',
      stop: 'Panama St. & Via Ortega',
      location: 'Stanford'
    },

    {
      hack: true
    },

    {
      day: 'Sunday, Februrary 17th',
      time: '3:30pm',
      stop: 'Check in at Panama St. & Via Ortega',
      location: 'Stanford'
    },

    {
      day: 'Sunday, Februrary 17th',
      time: '11:00pm',
      stop: 'UCI',
      location: 'Irvine'
    }
  ],
  [TRANSPORTATION_BUS_ROUTES.POMONA]: [
    {
      day: 'Friday, Februrary 15th',
      time: '10:45am',
      stop: 'Check in at Harvey Mudd (location added later)',
      location: 'Claremont'
    },

    {
      day: 'Friday, Feburary 15th',
      time: '6:00pm',
      stop: 'Panama St. & Via Ortega',
      location: 'Stanford'
    },

    {
      hack: true
    },

    {
      day: 'Sunday, Februrary 17th',
      time: '3:30pm',
      stop: 'Check in at Panama St. & Via Ortega',
      location: 'Stanford'
    },

    {
      day: 'Sunday, Februrary 17th',
      time: '10:00pm',
      stop: 'Harvey Mudd',
      location: 'Claremont'
    }
  ],
  [TRANSPORTATION_BUS_ROUTES.BERKELEY]: [
    {
      day: 'Friday, Februrary 15th',
      time: '2:30pm',
      stop: 'Check in at West Circle / University Drive',
      location: 'Berkeley'
    },

    {
      day: 'Friday, Feburary 15th',
      time: '4:30pm',
      stop: 'Panama St. & Via Ortega',
      location: 'Stanford'
    },

    {
      hack: true
    },

    {
      day: 'Sunday, Februrary 17th',
      time: '3:30pm',
      stop: 'Check in at Panama St. & Via Ortega',
      location: 'Stanford'
    },

    {
      day: 'Sunday, Februrary 17th',
      time: '10:00pm',
      stop: 'West Circle / University Drive',
      location: 'Berkeley'
    }
  ]
};

export const HACKATHON_YEAR = 2019;
export const HACKATHON_DATE_RANGE = "February 15-17";