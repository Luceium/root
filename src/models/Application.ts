import mongoose from "mongoose";
import { Model, Schema } from "mongoose";
import { IApplication } from "./Application.d"
import applicationInfoSchema from "./applicationInfoSchema";
import adminInfoSchema from "./adminInfoSchema";
import reviewSchema from "./reviewSchema";
import { STATUS, TRANSPORTATION_STATUS } from "../constants";
import {values} from "lodash";
import transportationInfoSchema from "./transportationInfoSchema";
import s3FilePlugin from "../utils/file_plugin";

export const applicationSchema: Schema = new mongoose.Schema({
    // user id is _id.
    "_id": String,
    "forms": { // can only be modified by user/editors
        "application_info": applicationInfoSchema,
        "transportation": transportationInfoSchema
    },
    "admin_info": adminInfoSchema, // Only editable by admin.
    "reviews": [reviewSchema], // each review can only be modified by the reviewer who made it.
    "user": {
        "email": String
    },
    "status": {
        type: String,
        default: STATUS.INCOMPLETE,
        enumValues: values(STATUS)
    },
    "transportation_status": {
        type: String,
        default: null,
        enumValues: values(TRANSPORTATION_STATUS)
    },
    "type": {
        type: String,
        enumValues: ["is", "oos", "stanford"]
    },
    "location": {
        type: String,
        enumValues: ["Outside USA", "Alabama", "Alaska", "American Samoa", "Arizona", "Arkansas", "California", "Colorado", "Connecticut", "Delaware", "District Of Columbia", "Florida", "Georgia", "Guam", "Hawaii", "Idaho", "Illinois", "Indiana", "Iowa", "Kansas", "Kentucky", "Louisiana", "Maine", "Maryland", "Massachusetts", "Michigan", "Minnesota", "Mississippi", "Missouri", "Montana", "Nebraska", "Nevada", "New Hampshire", "New Jersey", "New Mexico", "New York", "North Carolina", "North Dakota", "Ohio", "Oklahoma", "Oregon", "Pennsylvania", "Rhode Island", "South Carolina", "South Dakota", "Tennessee", "Texas", "Utah", "Vermont", "Virginia", "Washington", "West Virginia", "Wisconsin", "Wyoming", "US Territories"]
    }
});

applicationSchema.plugin(s3FilePlugin);

const model: Model<IApplication> = mongoose.model("Application", applicationSchema);
export default model; 