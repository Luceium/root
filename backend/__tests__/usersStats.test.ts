import request from "supertest";
import app from "../index";
import Application from "../models/Application";
import { omit } from "lodash";
import { STATUS, TYPE, HACKATHON_YEAR_STRING } from '../constants';

jest.mock("../constants");

const _doc = {
    reviews: [],
    type: TYPE.OUT_OF_STATE,
    status: STATUS.INCOMPLETE,
    transportation_status: null,
    user: {
        email: 'test@treehacks'
    },
    year: HACKATHON_YEAR_STRING,
    forms: {
        application_info: {
            first_name: "fir",
            last_name: "test",
            phone: "test",
            dob: "test",
            gender: "test",
            race: ["test"],
            university: "test",
            graduation_year: "test",
            level_of_study: "test",
            major: "test",
            skill_level: 1,
            hackathon_experience: 2,
            resume: "testtesttest",
            accept_terms: true,
            accept_share: true,
            q1: "test",
            q2: "test",
            q3: "test",
            q4: "test",
            q5: "test"
        }
    }
};

let docs = [
    { ..._doc, user: { email: 'test@treehacks', id: 'applicanttreehacks'} },
    { ..._doc, user: { email: 'test@treehacks', id: 'applicant-optout-confirmed'}, sponsor_optout: true, status: STATUS.ADMISSION_CONFIRMED },
    { ..._doc, user: { email: 'test@treehacks', id: 'applicant-confirmed'}, status: STATUS.ADMISSION_CONFIRMED },
    { ..._doc, user: { email: 'test@treehacks', id: 'applicant-confirmed-2'}, status: STATUS.ADMISSION_CONFIRMED },
    { ..._doc, user: { email: 'test@treehacks', id: 'applicant-submitted'}, status: STATUS.SUBMITTED },
    {
        ..._doc, user: { email: 'test@treehacks', id: 'applicant-name-thomas'},
        forms: {
            ..._doc.forms, application_info: {
                ..._doc.forms.application_info,
                first_name: 'thomas'
            }
        }
    },
    {
        ..._doc, user: { email: 'test@treehacks', id: 'applicant-name-tracey'},
        forms: {
            ..._doc.forms, application_info: {
                ..._doc.forms.application_info,
                first_name: 'tracey'
            }
        }
    }
];

describe('user stats by applicant', () => {
    test('view user stats - fail', () => {
        return request(app)
            .get("/api/users_stats")
            .set({ Authorization: 'applicant' })
            .expect(403);
    });
});

describe('user stats by sponsor', () => {
    test('view user stats - fail', () => {
        return request(app)
            .get("/api/users_stats")
            .set({ Authorization: 'applicant' })
            .expect(403);
    });
});

describe('user stats by admin', () => {
    test('view user stats', async () => {
        return request(app)
            .get("/api/users_stats")
            .set({ Authorization: 'admin' })
            .expect(200).then(e => {
                let body = e.body;
                for (let key in body) { // Date is not constant. TODO: Fix this
                    body[key] = omit(body[key], "date");
                }
                expect(body).toMatchSnapshot();
            });
    });
});