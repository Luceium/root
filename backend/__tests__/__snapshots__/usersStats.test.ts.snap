// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`user stats by admin view user stats 1`] = `
Object {
  "gender": Array [
    Object {
      "_id": "test",
      "count": 7,
    },
  ],
  "hackathon_experience": Array [
    Object {
      "_id": 2,
      "count": 7,
    },
  ],
  "location": Array [
    Object {
      "_id": null,
      "count": 7,
    },
  ],
  "race": Array [
    Object {
      "_id": "test",
      "count": 7,
    },
  ],
  "skill_level": Array [
    Object {
      "_id": 1,
      "count": 7,
    },
  ],
  "status": Array [
    Object {
      "_id": "admission_confirmed",
      "count": 3,
    },
    Object {
      "_id": "incomplete",
      "count": 3,
    },
    Object {
      "_id": "submitted",
      "count": 1,
    },
  ],
  "timeline": Array [
    Object {
      "date": "2019-11-04T23:55:34.000Z",
      "num_is": 0,
      "num_oos": 1,
      "num_stanford": 0,
    },
    Object {
      "date": "2019-11-04T23:55:34.000Z",
      "num_is": 0,
      "num_oos": 2,
      "num_stanford": 0,
    },
    Object {
      "date": "2019-11-04T23:55:34.000Z",
      "num_is": 0,
      "num_oos": 3,
      "num_stanford": 0,
    },
    Object {
      "date": "2019-11-04T23:55:34.000Z",
      "num_is": 0,
      "num_oos": 4,
      "num_stanford": 0,
    },
    Object {
      "date": "2019-11-04T23:55:34.000Z",
      "num_is": 0,
      "num_oos": 5,
      "num_stanford": 0,
    },
    Object {
      "date": "2019-11-04T23:55:34.000Z",
      "num_is": 0,
      "num_oos": 6,
      "num_stanford": 0,
    },
    Object {
      "date": "2019-11-04T23:55:34.000Z",
      "num_is": 0,
      "num_oos": 7,
      "num_stanford": 0,
    },
  ],
  "type": Array [
    Object {
      "_id": "oos",
      "count": 7,
    },
  ],
  "university": Array [
    Object {
      "_id": "test",
      "count": 7,
    },
  ],
}
`;
